executor/table_reader.go

```diff
@@ -0,0 +1,187 @@
+// Copyright 2017 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package executor
+
+import (
+	"github.com/juju/errors"
+	"github.com/pingcap/tidb/distsql"
+	"github.com/pingcap/tidb/model"
+	"github.com/pingcap/tidb/plan"
+	"github.com/pingcap/tidb/statistics"
+	"github.com/pingcap/tidb/table"
+	"github.com/pingcap/tidb/util/chunk"
+	"github.com/pingcap/tidb/util/ranger"
+	tipb "github.com/pingcap/tipb/go-tipb"
+	"golang.org/x/net/context"
+)
+
+// TableReaderExecutor sends DAG request and reads table data from kv layer.
+type TableReaderExecutor struct {
+	baseExecutor
+
+	table     table.Table
+	tableID   int64
+	keepOrder bool
+	desc      bool
+	ranges    []*ranger.Range
+	dagPB     *tipb.DAGRequest
+	// columns are only required by union scan.
+	columns []*model.ColumnInfo
+
+	// resultHandler handles the order of the result. Since (MAXInt64, MAXUint64] stores before [0, MaxInt64] physically
+	// for unsigned int.
+	resultHandler *tableResultHandler
+	streaming     bool
+	feedback      *statistics.QueryFeedback
+
+	// corColInFilter tells whether there's correlated column in filter.
+	corColInFilter bool
+	// corColInAccess tells whether there's correlated column in access conditions.
+	corColInAccess bool
+	plans          []plan.PhysicalPlan
+}
+
+// Open initialzes necessary variables for using this executor.
+func (e *TableReaderExecutor) Open(ctx context.Context) error {
+	span, ctx := startSpanFollowsContext(ctx, "executor.TableReader.Open")
+	defer span.Finish()
+
+	var err error
+	if e.corColInFilter {
+		e.dagPB.Executors, _, err = constructDistExec(e.ctx, e.plans)
+		if err != nil {
+			return errors.Trace(err)
+		}
+	}
+	if e.corColInAccess {
+		ts := e.plans[0].(*plan.PhysicalTableScan)
+		access := ts.AccessCondition
+		pkTP := ts.Table.GetPkColInfo().FieldType
+		e.ranges, err = ranger.BuildTableRange(access, e.ctx.GetSessionVars().StmtCtx, &pkTP)
+		if err != nil {
+			return errors.Trace(err)
+		}
+	}
+
+	e.resultHandler = &tableResultHandler{}
+	firstPartRanges, secondPartRanges := splitRanges(e.ranges, e.keepOrder)
```

Comment:
The function `splitRanges` and `tableResultHandler` should be moved too, it is used for the table reader.

Reply:
DONE

Reply:
But it's also used in index reader.

Reply:
So put into a util_reader.go?